rpc/RPCTransaction.java

```diff
@@ -23,82 +23,86 @@
 
-            final StreamObserver<TransactionProto.Transaction.Res> responseObserver = createResponseObserver();
-            requestObserver = session.getAsyncGrpcStub().transaction(responseObserver);
+            // Opening the StreamObserver exposes these atomics to another thread, so we must initialize them first.
+            streamIsOpen = new AtomicBoolean(false);
+            transactionWasOpened = new AtomicBoolean(false);
+            transactionWasClosed = new AtomicBoolean(false);
+            requestObserver = GraknGrpc.newStub(session.channel()).transaction(responseObserver());
+            streamIsOpen.set(true);
 
-            final TransactionProto.Transaction.Req openRequest = TransactionProto.Transaction.Req.newBuilder()
+            final TransactionProto.Transaction.Req.Builder openRequest = TransactionProto.Transaction.Req.newBuilder()
                     .putAllMetadata(tracingData())
                     .setOpenReq(TransactionProto.Transaction.Open.Req.newBuilder()
-                            .setSessionID(sessionId)
+                            .setSessionId(sessionId)
                             .setType(TransactionProto.Transaction.Type.forNumber(type.id()))
-                            .setOptions(options(options))).build();
-            execute(openRequest);
-            isOpen = new AtomicBoolean(true);
+                            .setOptions(options(options)));
+            final Instant startTime = Instant.now();
+            final TransactionProto.Transaction.Open.Res res = execute(openRequest).getOpenRes();
+            final Instant endTime = Instant.now();
+            transactionWasOpened.set(true);
+            networkLatencyMillis = (int) ChronoUnit.MILLIS.between(startTime, endTime) - res.getProcessingTimeMillis();
         }
     }
 
-    private StreamObserver<TransactionProto.Transaction.Res> createResponseObserver() {
-        return new StreamObserver<TransactionProto.Transaction.Res>() {
-            @Override
-            public synchronized void onNext(final TransactionProto.Transaction.Res value) {
-                processResponse(new RPCResponse.Ok(value));
-            }
-
-            @Override
-            public void onError(final Throwable throwable) {
-                assert throwable instanceof StatusRuntimeException : "The server only yields these exceptions";
-                while (collectorQueue.peek() != null)
-                    processResponse(new RPCResponse.Error((StatusRuntimeException) throwable));
-            }
-
-            @Override
-            public void onCompleted() {
-                while (collectorQueue.peek() != null) processResponse(new RPCResponse.Completed());
-            }
-        };
-    }
-
     @Override
     public Type type() {
         return type;
     }
 
     @Override
     public boolean isOpen() {
-        return isOpen.get();
+        return transactionWasOpened.get() && !transactionWasClosed.get();
```

Comment:
Technically nothing prevents open and close being triggered out of order, so we record both events.